### DISCUSSION TITLE
Welcome to Denovo Discussions!

### DISCUSSION THREAD

**Discussion by Coding-Enthusiast** [Autarkysoft/Denovo#3] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by AngeloMetal** [comment]
It seems that I'm the first one that is too excited about this project. I'm relatively a beginner on visual studio, so I would like to ask some questions about Bitcoin.net and the way nodes send information. On bitcointalk you said that commands like "getblockchaininfo" are something that happen locally and that cannot be received from a node. You mentioned that I'll have to use my local instance of blockchain, so I'll have to have the entire blockchain downloaded.

What I want to understand is if bitcoin.net provides sending and receiving information from network's nodes, without having a node. For example, I imagine running a program in which I firstly create a handshake with a node and then I send it to run me a command like "getbestblockhash". Once it does, it will return me the result (on JSON for example). Is this possible?

**Reply by Coding-Enthusiast** [reply]
> What I want to understand is if bitcoin.net provides sending and receiving information from network's nodes, without having a node.  

We are trying to create a stand-alone full node instead of relying on another application. The term "client" is used here to refer to the node you run and the term "node" is used for any other peer that you connect to.

>  I firstly create a handshake with a node and then I send it to run me a command like "getbestblockhash". Once it does, it will return me the result (on JSON for example). Is this possible?

What you want is not directly defined in bitcoin's [communication protocol](https://developer.bitcoin.org/reference/p2p_networking.html). Other nodes will tell you their best height as a number (in version message) which is not a reliable thing since it is not used anywhere; and they can respond with a list of block hashes but _only if_ you send them your own list of block hashes (that you store locally) first.  
For example imagine if there were 10 bitcoin blocks in total and you had stored 5 of them, you send 5 hashes to the other node and they check it against their own local blockchain if it matched then they will send you headers of block 6 to 10. Then by knowing these headers you can download the actual blocks too (they can also send you the blocks directly but that is the old way).  

A method such as "getbestblockhash" can not exist in a decentralized network because the node you connect to can return anything and you will have no way of verifying it.

**Reply by AngeloMetal** [reply]
Well, there are some ways to verify if a node is telling the truth or not. I could run the getbestblockhash command from 2 (or more) nodes at the same time. If both of them return the same result, then it's ensuring me that they're telling the truth. It doesn't mean that it's verified, but it's more ensuring.

I don't care that much about the verifying part. I care if it's possible to do this currently. You're telling me that it is not defined in bitcoin's communication protocol, but you also told me that they'll give me their best height as a number, it just won't be reliable. Is it possible or not? I didn't understand.

What I do want to achieve is a console in which you can get blockchain's information without having to carry it on your computer. Something similar must exist, but it doesn't bother me. It's a nice exercise.

**Reply by Coding-Enthusiast** [reply]
> I care if it's possible to do this currently.  

No, it is not possible to get the best hash.

> but you also told me that they'll give me their best height as a number, it just won't be reliable. Is it possible or not? I didn't understand.

When you connect to a node for the first time and receive their version message (during handshake) it will contain a number telling you their best chain's height. For example right now the version messages contain `661345` telling use the synced nodes have 661345 blocks verified and saved.

> What I do want to achieve is a console in which you can get blockchain's information without having to carry it on your computer. Something similar must exist, but it doesn't bother me. It's a nice exercise.

You'll have to rely on centralized services for that, such as block explorers through their API.

**Reply by AngeloMetal** [reply]
> You'll have to rely on centralized services for that, such as block explorers through their API.

But isn't block explorers simple nodes? I mean that by giving me their API, I'm telling them to run me commands and return me the result. I guess it's not the same, with connecting to nodes just with their IP address.

I had wrong view about nodes. I though that you could somehow get information from a node. Not just to receive blocks from it.

**Reply by Coding-Enthusiast** [reply]
> But isn't block explorers simple nodes?

Block explorers are a lot more _features_ built on top of a simple node.

> I had wrong view about nodes. I though that you could somehow get information from a node. Not just to receive blocks from it.

Well the purpose of a node is to validate blocks and transactions with optional propagation of them while enforcing rules. Anything more will be extra and certain things go against the decentralized and "verify; don't trust" principles.

**Comment by mraunt** [comment]
Hi how can i speak to you privatly? its emergancy for me. i have the wallet to recover it bu t idont know the code please help me/ my email is n.salati70@gmail.com please email to me
